# Stack Overflow Survey Analysis 

https://github.com/gkhayes/ds_survey_analysis/blob/master/DS%20Survey%20Analysis.ipynb

In [1]:
import pandas as pd
import numpy as np
import plotly as py
import plotly.figure_factory as ff

## 1. Overview
I look forward to reading Stack Overflow's annual Engineering Survey which is the largest and most comprehensive survey of people who code around the world. Every Year, Stack Overflow is providing us with many interesting insights.

If you are managing Stack Overflow website however, it is important to understand who are the actual users and the needs of the users in order to serve the users better. 

In this notebook, we will provide some insights on the survey by answering the questions below.

## 2. Research Questions
The motivation behind this analysis is to better understand the users of Stack Overflow in order to serve the users better. Conswquently, we set out to answer the following questions: 

1. Who are answering to Stack Overflow Survey? Are professionals or non-professionals using the site more? 
1. Who are answering to Stack Overflow Survey in terms of country? 
1. What is the most popular programming language? 
1. Is there any gap between programming language that people want to learn and what they're actually using?
1. Is it true that developers have all been programming since childhood?

## 3. Data Understanding

In order to answer these questions, we make use of data collected by Stack Overflow in response to their 2017 Annual Developr Survey. The data can be found [here](https://www.kaggle.com/stackoverflow/so-survey-2017).

In [2]:
survey = pd.read_csv("./survey_results_public.csv")
survey.head()

Respondent                                       Professional  \
0           1                                            Student   
1           2                                            Student   
2           3                             Professional developer   
3           4  Professional non-developer who sometimes write...   
4           5                             Professional developer   

                ProgramHobby         Country      University  \
0                  Yes, both   United States              No   
1                  Yes, both  United Kingdom  Yes, full-time   
2                  Yes, both  United Kingdom              No   
3                  Yes, both   United States              No   
4  Yes, I program as a hobby     Switzerland              No   

                         EmploymentStatus  \
0  Not employed, and not looking for work   
1                      Employed part-time   
2                      Employed full-time   
3                      Employed full-time   
4                      Employed full-time   

                                     FormalEducation  \
0                                   Secondary school   
1  Some college/university study without earning ...   
2                                  Bachelor's degree   
3                                    Doctoral degree   
4                                    Master's degree   

                                  MajorUndergrad  \
0                                            NaN   
1       Computer science or software engineering   
2       Computer science or software engineering   
3  A non-computer-focused engineering discipline   
4       Computer science or software engineering   

                                          HomeRemote  \
0                                                NaN   
1              More than half, but not all, the time   
2  Less than half the time, but at least one day ...   
3  Less than half the time, but at least one day ...   
4                                              Never   

                CompanySize       ...       StackOverflowMakeMoney Gender  \
0                       NaN       ...            Strongly disagree   Male   
1        20 to 99 employees       ...            Strongly disagree   Male   
2  10,000 or more employees       ...                     Disagree   Male   
3  10,000 or more employees       ...                     Disagree   Male   
4        10 to 19 employees       ...                          NaN    NaN   

  HighestEducationParents                          Race         SurveyLong  \
0             High school  White or of European descent  Strongly disagree   
1       A master's degree  White or of European descent     Somewhat agree   
2   A professional degree  White or of European descent     Somewhat agree   
3       A doctoral degree  White or of European descent              Agree   
4                     NaN                           NaN                NaN   

  QuestionsInteresting QuestionsConfusing InterestedAnswers    Salary  \
0       Strongly agree           Disagree    Strongly agree       NaN   
1       Somewhat agree           Disagree    Strongly agree       NaN   
2                Agree           Disagree             Agree  113750.0   
3                Agree     Somewhat agree    Strongly agree       NaN   
4                  NaN                NaN               NaN       NaN   

   ExpectedSalary  
0             NaN  
1         37500.0  
2             NaN  
3             NaN  
4             NaN  

[5 rows x 154 columns]

### Size of Data

In [3]:
# Print shape of dataset
print('The dataset contains', np.shape(survey)[0], 'rows and', np.shape(survey)[1], 'columns.')

The dataset contains 51392 rows and 154 columns.


### Available Columns

In [4]:
# Print column names
list(survey.columns.values)

['Respondent',
 'Professional',
 'ProgramHobby',
 'Country',
 'University',
 'EmploymentStatus',
 'FormalEducation',
 'MajorUndergrad',
 'HomeRemote',
 'CompanySize',
 'CompanyType',
 'YearsProgram',
 'YearsCodedJob',
 'YearsCodedJobPast',
 'DeveloperType',
 'WebDeveloperType',
 'MobileDeveloperType',
 'NonDeveloperType',
 'CareerSatisfaction',
 'JobSatisfaction',
 'ExCoderReturn',
 'ExCoderNotForMe',
 'ExCoderBalance',
 'ExCoder10Years',
 'ExCoderBelonged',
 'ExCoderSkills',
 'ExCoderWillNotCode',
 'ExCoderActive',
 'PronounceGIF',
 'ProblemSolving',
 'BuildingThings',
 'LearningNewTech',
 'BoringDetails',
 'JobSecurity',
 'DiversityImportant',
 'AnnoyingUI',
 'FriendsDevelopers',
 'RightWrongWay',
 'UnderstandComputers',
 'SeriousWork',
 'InvestTimeTools',
 'WorkPayCare',
 'KinshipDevelopers',
 'ChallengeMyself',
 'CompetePeers',
 'ChangeWorld',
 'JobSeekingStatus',
 'HoursPerWeek',
 'LastNewJob',
 'AssessJobIndustry',
 'AssessJobRole',
 'AssessJobExp',
 'AssessJobDept',
 'AssessJobT

In [5]:
# Drop unnecessary columns
survey = survey[['Respondent','Professional', 'Gender', 'Country', 'HaveWorkedLanguage' ,'WantWorkLanguage','YearsProgram','YearsCodedJob']]

In [6]:
# Summary of the data
survey.drop(['Respondent'], axis = 1).describe(include = 'all')

Professional Gender        Country   HaveWorkedLanguage  \
count                    51392  35047          51392                36625   
unique                       5     29            201                 8438   
top     Professional developer   Male  United States  C#; JavaScript; SQL   
freq                     36131  31589          11455                 1276   

       WantWorkLanguage      YearsProgram YearsCodedJob  
count             33771             51145         40890  
unique            11239                21            21  
top          JavaScript  20 or more years  1 to 2 years  
freq                642              8817          5289

### Let's look at unique values for each of these features

In [9]:
survey['Professional'].value_counts()

Professional developer                                  36131
Student                                                  8224
Professional non-developer who sometimes writes code     5140
Used to be a professional developer                       983
None of these                                             914
Name: Professional, dtype: int64

In [75]:
survey['Gender'].value_counts()

Male      31589
Others    17203
Female     2600
Name: Gender, dtype: int64

In [79]:
survey['Country'].value_counts().head(20)

United States         11455
India                  5197
United Kingdom         4395
Germany                4143
Canada                 2233
France                 1740
Poland                 1290
Australia               913
Russian Federation      873
Spain                   864
Netherlands             855
Italy                   781
Brazil                  777
Sweden                  611
Switzerland             595
Israel                  575
Romania                 561
Iran                    507
Austria                 477
Pakistan                454
Name: Country, dtype: int64

In [11]:
survey['YearsProgram'].value_counts()

20 or more years    8817
4 to 5 years        3865
3 to 4 years        3701
5 to 6 years        3572
2 to 3 years        3251
9 to 10 years       3197
6 to 7 years        2869
1 to 2 years        2773
7 to 8 years        2466
10 to 11 years      2178
14 to 15 years      2011
8 to 9 years        1910
15 to 16 years      1669
Less than a year    1484
11 to 12 years      1389
12 to 13 years      1307
13 to 14 years      1092
16 to 17 years      1046
19 to 20 years      1021
17 to 18 years       879
18 to 19 years       648
Name: YearsProgram, dtype: int64

In [12]:
survey['HaveWorkedLanguage'].value_counts().nlargest(20)

C#; JavaScript; SQL                1276
JavaScript; PHP; SQL               1143
Java                                913
JavaScript                          807
JavaScript; PHP                     662
Java; JavaScript; SQL               645
Java; JavaScript                    585
C#                                  540
Python                              529
C#; JavaScript                      500
C#; SQL                             494
C#; JavaScript; SQL; TypeScript     410
JavaScript; Python                  384
Java; SQL                           356
Java; JavaScript; PHP; SQL          342
C#; JavaScript; PHP; SQL            256
JavaScript; PHP; Python; SQL        253
C++                                 247
C#; Java; JavaScript; SQL           230
PHP                                 217
Name: HaveWorkedLanguage, dtype: int64

In [13]:
survey['WantWorkLanguage'].value_counts().nlargest(10)

JavaScript                         642
Python                             556
Java                               544
C#                                 475
C#; JavaScript; SQL                444
Java; JavaScript                   332
C#; JavaScript; SQL; TypeScript    326
JavaScript; PHP; SQL               324
C++                                317
Swift                              306
Name: WantWorkLanguage, dtype: int64

## 4. Data Preparation

### Simplify Data

#### Simplify Gender to Male, Female, and Others categories

In [14]:
survey['Gender'] = survey['Gender'].where(survey['Gender'].isin(['Male','Female']),'Others')

#### Transform Years coded job into float numbers

In [17]:
dict = {}
i = 1
while i < 20:
    dict["{} to {} years".format(i,i+1)] = i+0.5
    i+=1
dict["20 or more years"] = 21
dict["Less than a year"] = 0.5
column = "YearsCodedJob"
survey[column] = survey[column].replace(dict)
survey[column].value_counts()
# survey[column] = survey[column].where(survey[column].isin(dict.values()),'Others')

1.5     5289
2.5     4767
3.5     4003
4.5     3405
21.0    3067
0.5     3034
5.5     2990
9.5     1956
6.5     1912
10.5    1678
7.5     1638
8.5     1287
15.5     845
14.5     840
11.5     838
12.5     741
16.5     694
17.5     545
13.5     531
19.5     427
18.5     403
Name: YearsCodedJob, dtype: int64

#### Transform Years Program into float numbers

In [18]:
dict = {}
i = 1
while i < 20:
    dict["{} to {} years".format(i,i+1)] = i+0.5
    i+=1
dict["20 or more years"] = 21
dict["Less than a year"] = 0.5
column = "YearsProgram"
survey[column] = survey[column].replace(dict)
survey[column].value_counts()
# survey[column] = survey[column].where(survey[column].isin(dict.values()),'Others')

21.0    8817
4.5     3865
3.5     3701
5.5     3572
2.5     3251
9.5     3197
6.5     2869
1.5     2773
7.5     2466
10.5    2178
14.5    2011
8.5     1910
15.5    1669
0.5     1484
11.5    1389
12.5    1307
13.5    1092
16.5    1046
19.5    1021
17.5     879
18.5     648
Name: YearsProgram, dtype: int64

## 5. Analysis, Modeling, Visualization

### Question 1: Who are answering to Stack Overflow Survey? Are professionals or non-professionals using the site more?
### Graph 1-1: Professional by Gender (Absolute Value)

In [97]:
import plotly.graph_objects as go
data = survey[['YearsProgram','YearsCodedJob','Gender','Professional','Country']].dropna().copy()
fig = go.Figure()


print('Total Number of Male  : {}   {}%'.format(len(data[data['Gender']=='Male']["Professional"]),100*len(data[data['Gender']=='Male']["Professional"])/len(data)))
print('Total Number of Female: {}   {}%'.format(len(data[data['Gender']=='Female']["Professional"]),100*len(data[data['Gender']=='Female']["Professional"])/len(data)))
print('Total Number of Others: {}   {}%'.format(len(data[data['Gender']=='Others']["Professional"]),100*len(data[data['Gender']=='Others']["Professional"])/len(data)))

fig.add_trace(go.Histogram(x=data[data['Gender']=='Male']["Professional"],name = 'Male'))
fig.add_trace(go.Histogram(x=data[data['Gender']=='Female']["Professional"],name = 'Female'))
fig.add_trace(go.Histogram(x=data[data['Gender']=='Others']["Professional"],name = 'Others'))
fig.update_layout(barmode='group')

# fig = px.histogram(data, x="YearsProgram",color="Gender",marginal="rug")
# fig.update_xaxes(range=[0, 21])
fig.show()

Total Number of Male  : 26477   64.80248666111899%
Total Number of Female: 2152   5.267022370160067%
Total Number of Others: 12229   29.930490968720935%


### Graph 1-2: Professional by Gender (%)

In [109]:
data_professional_percentage = pd.get_dummies(data[['Professional','Gender']], columns = ['Professional'],prefix = '',prefix_sep='').groupby('Gender').sum().stack()

data_professional_percentage['Male'] = data_professional_percentage['Male']/data_professional_percentage['Male'].sum()
data_professional_percentage['Female'] = data_professional_percentage['Female']/data_professional_percentage['Female'].sum()
data_professional_percentage['Others'] = data_professional_percentage['Others']/data_professional_percentage['Others'].sum()

data_professional_percentage = data_professional_percentage.reset_index().rename(columns={'level_1':'Professional',0:"Percentage"})
fig = px.bar(data_professional_percentage, x='Professional', y='Percentage',barmode='group',color = 'Gender')
fig.show()

data_professional_percentage

Gender                                       Professional  Percentage
0  Female                             Professional developer    0.848978
1  Female  Professional non-developer who sometimes write...    0.151022
2  Female                                            Student    0.000000
3    Male                             Professional developer    0.892548
4    Male  Professional non-developer who sometimes write...    0.107301
5    Male                                            Student    0.000151
6  Others                             Professional developer    0.845531
7  Others  Professional non-developer who sometimes write...    0.154469
8  Others                                            Student    0.000000

### Insight 1: Most of the repondents are professionals (more than 84% for all the genders) and mostly Male (64.8%)

### Question 2: Who are answering to Stack Overflow Survey in terms of country?
### Grapht 2-1: Number of Male Respondents by country

In [132]:
import plotly.express as px
### Male
geo = survey[survey['Gender']=='Male'].groupby('Country').count().reset_index().copy()
geo['Respondent_percent'] = geo['Respondent']/geo['Respondent'].sum()
print('Male: ',geo[['Country','Respondent','Respondent_percent']].sort_values(by = ['Respondent'],ascending = False).head(10).reset_index(drop = True))
fig = px.scatter_geo(geo, locations = 'Country',size = 'Respondent',hover_name="Country",locationmode = 'country names',
                     projection="natural earth")
fig.show()

Male:                Country  Respondent  Respondent_percent
0       United States        7447            0.235747
1      United Kingdom        2954            0.093514
2             Germany        2740            0.086739
3               India        2422            0.076672
4              Canada        1446            0.045775
5              France        1094            0.034632
6              Poland         806            0.025515
7           Australia         634            0.020070
8         Netherlands         598            0.018931
9  Russian Federation         542            0.017158


### Grapht 2-2: Number of Female Respondents by country

In [131]:
### Female
geo = survey[survey['Gender']=='Female'].groupby('Country').count().reset_index().copy()
geo['Respondent_percent'] = geo['Respondent']/geo['Respondent'].sum()
print('Female: ',geo[['Country','Respondent','Respondent_percent']].sort_values(by = ['Respondent'],ascending = False).head(10).reset_index(drop = True))
fig = px.scatter_geo(geo, locations = 'Country',size = 'Respondent',hover_name="Country",locationmode = 'country names',
                     projection="natural earth")
fig.show()

Female:            Country  Respondent  Respondent_percent
0   United States         847            0.325769
1           India         230            0.088462
2  United Kingdom         221            0.085000
3         Germany         153            0.058846
4          Canada         150            0.057692
5          Poland          98            0.037692
6          France          64            0.024615
7           Spain          41            0.015769
8          Israel          40            0.015385
9       Australia          40            0.015385


### Grapht 2-3: Number of 'Others' Respondents by country

In [133]:
### Others
geo = survey[survey['Gender']=='Others'].groupby('Country').count().reset_index().copy()
geo['Respondent_percent'] = geo['Respondent']/geo['Respondent'].sum()
print('Others: ',geo[['Country','Respondent','Respondent_percent']].sort_values(by = ['Respondent'],ascending = False).head(10).reset_index(drop = True))
fig = px.scatter_geo(geo, locations = 'Country',size = 'Respondent',hover_name="Country",locationmode = 'country names',
                     projection="natural earth")
fig.show()

Others:                Country  Respondent  Respondent_percent
0       United States        3161            0.183747
1               India        2545            0.147939
2             Germany        1250            0.072662
3      United Kingdom        1220            0.070918
4              Canada         637            0.037028
5              France         582            0.033831
6              Poland         386            0.022438
7               Spain         305            0.017729
8  Russian Federation         300            0.017439
9              Brazil         261            0.015172


## Result 2: Have Worked vs Want to Work

In [22]:
column = 'WantWorkLanguage'
dummies = survey[column].str.get_dummies(sep="; ").add_prefix(column+'_')
unique_WantWorkLanguage = dummies.columns
survey = pd.concat([survey.drop(column, 1), dummies], 1)

In [23]:
survey

Respondent                                       Professional  Gender  \
0               1                                            Student    Male   
1               2                                            Student    Male   
2               3                             Professional developer    Male   
3               4  Professional non-developer who sometimes write...    Male   
4               5                             Professional developer  Others   
5               6                                            Student  Others   
6               7  Professional non-developer who sometimes write...    Male   
7               8                             Professional developer    Male   
8               9                             Professional developer    Male   
9              10                             Professional developer    Male   
10             11  Professional non-developer who sometimes write...  Female   
11             12                             Professional developer  Others   
12             13                Used to be a professional developer    Male   
13             14                             Professional developer  Female   
14             15                             Professional developer    Male   
15             16                             Professional developer    Male   
16             17                             Professional developer    Male   
17             18                             Professional developer    Male   
18             19                             Professional developer  Others   
19             20                             Professional developer    Male   
20             21                             Professional developer  Others   
21             22                             Professional developer    Male   
22             23                             Professional developer    Male   
23             24                             Professional developer    Male   
24             25                             Professional developer  Others   
25             26                             Professional developer    Male   
26             27                             Professional developer  Others   
27             28                             Professional developer  Others   
28             29  Professional non-developer who sometimes write...    Male   
29             30  Professional non-developer who sometimes write...  Others   
...           ...                                                ...     ...   
51362       51363                             Professional developer    Male   
51363       51364                             Professional developer    Male   
51364       51365                             Professional developer    Male   
51365       51366                             Professional developer    Male   
51366       51367                             Professional developer  Others   
51367       51368                                            Student  Female   
51368       51369                             Professional developer    Male   
51369       51370                                            Student    Male   
51370       51371                             Professional developer    Male   
51371       51372                             Professional developer    Male   
51372       51373                             Professional developer  Others   
51373       51374                             Professional developer    Male   
51374       51375                             Professional developer  Others   
51375       51376                             Professional developer    Male   
51376       51377                                            Student    Male   
51377       51378                             Professional developer    Male   
51378       51379                             Professional developer    Male   
51379       51380                                            Student  Others   
51380  

In [24]:
survey[unique_WantWorkLanguage].sum(axis = 0).nlargest(20)

WantWorkLanguage_JavaScript     16327
WantWorkLanguage_Python         12570
WantWorkLanguage_SQL            11217
WantWorkLanguage_C#             10187
WantWorkLanguage_Java            9988
WantWorkLanguage_C++             7647
WantWorkLanguage_Go              5770
WantWorkLanguage_TypeScript      5692
WantWorkLanguage_PHP             5365
WantWorkLanguage_C               4841
WantWorkLanguage_Swift           4530
WantWorkLanguage_Ruby            3966
WantWorkLanguage_Scala           3122
WantWorkLanguage_Rust            2718
WantWorkLanguage_R               2628
WantWorkLanguage_Haskell         2306
WantWorkLanguage_F#              2034
WantWorkLanguage_Assembly        1923
WantWorkLanguage_Objective-C     1844
WantWorkLanguage_Elixir          1565
dtype: int64

In [25]:
column = 'HaveWorkedLanguage'
dummies = survey[column].str.get_dummies(sep="; ").add_prefix(column+'_')
unique_HaveWorkedLanguage = dummies.columns
survey = pd.concat([survey.drop(column, 1), dummies], 1)

In [26]:
data_haveworked=survey[unique_HaveWorkedLanguage].sum(axis = 0).copy()#.nlargest(20)
data_haveworked = data_haveworked.reset_index().rename(columns={'index':'HaveWorkedLanguage',0:"Percentage_HaveWorkedLanguage"})
data_haveworked['HaveWorkedLanguage'] = data_haveworked['HaveWorkedLanguage'].str.lstrip('HaveWorkedLanguage_')
data_haveworked['Percentage_HaveWorkedLanguage'] = data_haveworked['Percentage_HaveWorkedLanguage']/data_haveworked['Percentage_HaveWorkedLanguage'].sum()  
data_haveworked

HaveWorkedLanguage  Percentage_HaveWorkedLanguage
0            Assembly                       0.013147
1                   C                       0.050294
2                  C#                       0.089973
3                 C++                       0.058811
4             Clojure                       0.002820
5        CoffeeScript                       0.008596
6         Common Lisp                       0.001969
7                Dart                       0.001046
8              Elixir                       0.002740
9              Erlang                       0.002026
10                 F#                       0.003296
11                 Go                       0.011229
12             Groovy                       0.008604
13                 ck                       0.000772
14              skell                       0.004680
15               Java                       0.104742
16         JavaScript                       0.164967
17              Julia                       0.000995
18                                          0.007493
19             Matlab                       0.011315
20        Objective-C                       0.016940
21                PHP                       0.074208
22               Perl                       0.011431
23             Python                       0.084405
24                  R                       0.011784
25               Ruby                       0.023972
26               Rust                       0.003000
27                SQL                       0.135248
28              Scala                       0.009440
29          Smalltalk                       0.002358
30              Swift                       0.017077
31         TypeScript                       0.025154
32             VB.NET                       0.016392
33                VBA                       0.011351
34     Visual Basic 6                       0.007724

In [27]:
data_wantwork=survey[unique_WantWorkLanguage].sum(axis = 0)
data_wantwork = data_wantwork.reset_index().rename(columns={'index':'WantWorkLanguage',0:"Percentage_WantWorkLanguage"})
data_wantwork['WantWorkLanguage'] = data_wantwork['WantWorkLanguage'].str.lstrip('WantWorkLanguage_')
data_wantwork['Percentage_WantWorkLanguage'] = data_wantwork['Percentage_WantWorkLanguage']/data_wantwork['Percentage_WantWorkLanguage'].sum()
data_wantwork

WantWorkLanguage  Percentage_WantWorkLanguage
0          Assembly                     0.015071
1                 C                     0.037940
2                C#                     0.079837
3               C++                     0.059931
4           Clojure                     0.010251
5      CoffeeScript                     0.008088
6       Common Lisp                     0.005071
7              Dart                     0.003887
8            Elixir                     0.012265
9            Erlang                     0.008253
10               F#                     0.015941
11               Go                     0.045220
12           Groovy                     0.006787
13             Hack                     0.002649
14          Haskell                     0.018073
15             Java                     0.078278
16       JavaScript                     0.127958
17            Julia                     0.004036
18                                      0.008754
19           Matlab                     0.007296
20      Objective-C                     0.014452
21              PHP                     0.042046
22             Perl                     0.008535
23           Python                     0.098513
24                R                     0.020596
25             Ruby                     0.031082
26             Rust                     0.021301
27              SQL                     0.087910
28            Scala                     0.024468
29        Smalltalk                     0.003299
30            Swift                     0.035502
31       TypeScript                     0.044609
32           VB.NET                     0.006301
33              VBA                     0.003488
34   Visual Basic 6                     0.002312

In [28]:
data_haveworked.index = data_haveworked.HaveWorkedLanguage
data_wantwork.index = data_wantwork.WantWorkLanguage

In [29]:
import plotly.graph_objects as go
data_wantwork_haveworked = pd.concat([data_haveworked,data_wantwork],axis = 1).drop(columns = ['HaveWorkedLanguage','WantWorkLanguage']).fillna(0).reset_index().rename(columns={'index':'Language'})
data_wantwork_haveworked['WantWork-HaveWorked'] = data_wantwork_haveworked['Percentage_WantWorkLanguage'] - data_wantwork_haveworked['Percentage_HaveWorkedLanguage'] 
data_wantwork_haveworked = data_wantwork_haveworked.sort_values(by = ['WantWork-HaveWorked'],ascending = False)

fig = go.Figure()
fig.add_trace(go.Bar(x=data_wantwork_haveworked['Language'],y = data_wantwork_haveworked['WantWork-HaveWorked'],name = 'WantWork-HaveWorked' ))
fig.add_trace(go.Bar(x=data_wantwork_haveworked['Language'],y = data_wantwork_haveworked['Percentage_WantWorkLanguage'],name = 'Pecentage_WantWorkLanguage' ))
fig.add_trace(go.Bar(x=data_wantwork_haveworked['Language'],y = data_wantwork_haveworked['Percentage_HaveWorkedLanguage'] ,name = 'Percentage_HaveWorkedLanguage'))
# fig.add_trace(go.Histogram(x=data[data['Gender']=='Others']["Professional"]))
# fig.add_trace(go.Histogram(x=data[data['Gender']=='Female']["Professional"]))
fig.update_layout(barmode='group')
# fig = px.histogram(data, x="YearsProgram",color="Gender",marginal="rug")
# fig.update_xaxes(range=[0, 21])
# fig.show()
# 'Percentage_WantWorkLanguage'
fig.show()

## Result 3 Years Programming vs Years Coded Job

In [30]:
data['YearsProgram-YearsCodedJob'] = data['YearsProgram'] - data['YearsCodedJob'] 
tmp = data[['Gender','YearsProgram-YearsCodedJob','YearsCodedJob']]
tmp = tmp.groupby(['Gender','YearsCodedJob']).mean().reset_index()
fig = px.bar(tmp, x='YearsCodedJob', y='YearsProgram-YearsCodedJob',barmode='group',color = 'Gender')
fig.show()
tmp.groupby('Gender')['YearsProgram-YearsCodedJob'].mean()

Gender
Female    3.869736
Male      4.448740
Others    3.532413
Name: YearsProgram-YearsCodedJob, dtype: float64